src/features/completion.ts

```diff
@@ -972,3 +1062,184 @@ export function jsonCompletion(opts: JSONCompletionOptions = {}) {
     return completion.doComplete(ctx);
   };
 }
+
+/**
+ * removes required properties and allows additional properties everywhere
+ * @param schema
+ */
+function makeSchemaLax(schema: any): any {
+  return replacePropertiesDeeply(schema, (key, value) => {
+    if (key === "additionalProperties" && value === false) {
+      return [];
+    }
+    if (key === "required" && Array.isArray(value)) {
+      return [];
+    }
+    if (key === "unevaluatedProperties" && value === false) {
+      return [];
+    }
+    if (key === "unevaluatedItems" && value === false) {
+      return [];
+    }
+    // TODO remove dependencies and other restrictions
+    // if (key === 'dependencies' && typeof value === 'object') {
+    //   return Object.keys(value).reduce((acc: any, depKey) => {
+    //     const depValue = value[depKey];
+    //     if (Array.isArray(depValue)) {
+    //       return acc;
+    //     }
+    //     return { ...acc, [depKey]: depValue };
+    //   }, {});
+    // }
+    return [key, value];
+  });
+}
+
+/**
+ * determines effective object schema for given data
+ * TODO support patternProperties, etc.
+ * @param schema
+ * @param data
+ * @param pointer
+ */
+function getEffectiveObjectWithPropertiesSchema(
+  schema: JSONSchema7,
+  data: unknown,
+  pointer: string | undefined,
+): JSONSchema7 | undefined {
+  // TODO (unimportant): [performance] cache Draft07 in case it does some pre-processing? but does not seem to be significant
+  const draft = new Draft07(schema);
+  const subSchema = draft.getSchema({
+    pointer,
+    data: data ?? undefined,
+  });
+  if (!isRealSchema(subSchema)) {
+    return undefined;
+  }
+
+  const possibleDirectPropertyNames = getAllPossibleDirectStaticPropertyNames(
+    draft,
+    subSchema as JSONSchema7,
+  );
+  const effectiveProperties: Exclude<JSONSchema7["properties"], undefined> = {};
+  for (let possibleDirectPropertyName of possibleDirectPropertyNames) {
+    let propertyPointer = extendJsonPointer(
+      pointer,
+      possibleDirectPropertyName,
+    );
+    const subSchemaForPropertyConsideringData = draft.getSchema({
+      // TODO [performance] use subSchema and only check it's sub-properties
+      pointer: propertyPointer,
+      data: data ?? undefined,
+      // pointer: `/${possibleDirectPropertyName}`,
+      // schema: subSchema
+    });
+    if (isRealSchema(subSchemaForPropertyConsideringData)) {
+      Object.assign(effectiveProperties, {
+        [possibleDirectPropertyName]: subSchemaForPropertyConsideringData,
+      });
+    }
+  }
+
+  if (
+    possibleDirectPropertyNames.length === 0 ||
+    Object.keys(effectiveProperties).length === 0
+  ) {
+    // in case json-schema-library behaves too weirdly and returns nothing, just return no schema too to let other cases handle this edge-case
+    return undefined;
+  }
+
+  // TODO also resolve patternProperties of allOf, anyOf, oneOf
+  return {
+    ...omit(subSchema, ["allOf", "anyOf", "oneOf"]),
```

Comment:
there is an easier way to do this that doesn't require an additional dependency. radash looks cool but I don't think we need it right now, what do you think @imolorhe ?

a more es6 native/expressive way of doing this would be something more like this:

```ts
const {
 allOf, anyOf, oneOf, 
 ...simpleSchema
} = subSchema

return {
   ...simpleSchema,
  properties: effectiveProperties,
};
```

Reply:
Yeah we shouldn't be adding that extra dependency.